orca-core/src/test/groovy/com/netflix/spinnaker/orca/pipeline/persistence/ExecutionRepositoryTck.groovy

```diff
@@ -700,4 +705,78 @@ class JedisExecutionRepositorySpec extends ExecutionRepositoryTck<JedisExecution
       stages.type == ["one", "three"]
     }
   }
+
+  @Unroll
+  def "can add a synthetic stage #position"() {
+    given:
+    def pipeline = Pipeline
+      .builder()
+      .withApplication("orca")
+      .withName("dummy-pipeline")
+      .withStage("one")
+      .withStage("two")
+      .withStage("three")
+      .build()
+
+    repository.store(pipeline)
+
+    expect:
+    repository.retrievePipeline(pipeline.id).stages.size() == 3
+
+    when:
+    def stage = newStage(pipeline, "whatever", "two-whatever", [:], pipeline.namedStage("two"), position)
+    repository.addStage(stage)
+
+    then:
+    with(repository.retrievePipeline(pipeline.id)) {
+      stages.size() == 4
+      stages.name == expectedStageNames
+    }
+
+    where:
+    position     | expectedStageNames
+    STAGE_BEFORE | ["one", "two-whatever", "two", "three"]
+    STAGE_AFTER  | ["one", "two", "two-whatever", "three"]
+  }
+
+  def "can concurrently add stages without overwriting"() {
+    given:
+    def pipeline = Pipeline
+      .builder()
+      .withApplication("orca")
+      .withName("dummy-pipeline")
+      .withStage("one")
+      .withStage("two")
+      .withStage("three")
+      .build()
+
+    repository.store(pipeline)
+
+    expect:
+    repository.retrievePipeline(pipeline.id).stages.size() == 3
+
+    when:
+    def stage1 = newStage(pipeline, "whatever", "one-whatever", [:], pipeline.namedStage("one"), STAGE_BEFORE)
+    def stage2 = newStage(pipeline, "whatever", "two-whatever", [:], pipeline.namedStage("two"), STAGE_BEFORE)
+    def stage3 = newStage(pipeline, "whatever", "three-whatever", [:], pipeline.namedStage("three"), STAGE_BEFORE)
+    def startLatch = new CountDownLatch(1)
+    def doneLatch = new CountDownLatch(3)
+    [stage1, stage2, stage3].each { stage ->
+      Thread.start {
+        startLatch.await(1, SECONDS)
+        repository.addStage(stage)
+        doneLatch.countDown()
+      }
+    }
+    startLatch.countDown()
```

Comment:
This should prove we can add multiple stages to the same pipeline simultaneously without issues